docs/rfc_service_layer_document.md

```diff
@@ -0,0 +1,206 @@
+
+-   **Example of Current View for Disbanding (Deleting) a Team**:
+
+    ```python
+    class DisbandTeamAPIView(TeamPermissionsMixin, DestroyAPIView):
+        queryset = Team.objects.all()
+        lookup_url_kwarg = "team_name"
+        lookup_field = "name"
+
+        def check_permissions(self, request):
+            super().check_permissions(request)
+            team = self.get_object()
+            if not team.can_user_disband(request.user):
+                raise PermissionDenied("You do not have permission to disband this team.")
+
+        @conditional_swagger_auto_schema(
+            operation_id="cyberstorm.team.disband",
+            tags=["cyberstorm"],
+            responses={status.HTTP_204_NO_CONTENT: ""},
+        )
+        def delete(self, request, *args, **kwargs):
+            return super().delete(request, *args, **kwargs)
+    ```
+
+    The issue with this view is that it contains business logic related to deleting a team directly from the database. The view inherits from two other classes and performs permissions checking. The deletion logic and permissions checking could be moved elsewhere.
+
+## 2.2 Proposed Solution
+
+Move business logic from views to a separate service layer in each app with a clear domain. This will simplify views, making them easier to maintain, and will facilitate testing views and services in isolation.
+
+-   **Example of File Structure**:
+
+    ```
+    my_app/
+        ├── services/
+        │   ├── __init__.py
+        │   ├── team.py
+        ├── views/
+        │   ├── __init__.py
+        │   ├── team.py
+        └── tests/
+            ├── __init__.py
+            ├── test_services/
+            │   ├── __init__.py
+            │   ├── test_team.py
+            └── test_views/
+                ├── __init__.py
+                ├── test_team.py
+    ```
+
+# 3. Architecture Overview
+
+-   **Service Layer**: A new `services` module in each app will house files with functions containing core business logic (e.g., permissions checks, object manipulation). These functions will be small, focused on a single task, and will raise appropriate exceptions when necessary.
+
+-   **Views**: Views will be simplified to handle requests, responses, serialization, and the delegation of tasks.
+
+-   **Tests**: Tests will be updated to test the service layer in isolation, and views will be tested to ensure they handle requests and responses correctly.
+
+-   **Data Flow**:
+
+    1. The view receives a request.
+    2. The view delegates the business logic to the service layer.
+    3. The service layer performs the business logic, accesses the database, and returns a response.
+    4. The view handles the response and returns it to the client.
+
+    **Example**:
+
+    -   Request → View Layer (delegates) → Service Layer (accesses) → Database
+    -   Database → Service Layer (returns) → View Layer (returns) → Response
```

Comment:
View layer can probably be considered as formatting the response? Service layer returns a single known data format, views might mangle it as is needed by the view in question.